### PR TITLE
Google drive integration

### PR DESCRIPTION
**Pull request by ackagel**
## **Purpose**

Adds intuitive Google Drive support to ark-analysis.  Users will be able to clone and upload directory structures from/to Google Drive and their local filesystem.  Additionally, files can be directly read/written from/to Google Drive without saving a file locally.

## **Implementation**

Adds the `GoogleDrivePath` object and generalizes path manipulations to accommodate said GoogleDrivePaths.  Uses the Google Drive API which requires credentials which are encrypted via a secret key, and stored in a hidden directory.  The user must provide this secret key via the `init_google_drive_api()` function in order to use `GoogleDrivePath` objects.

## **Remaining issues**

Still need to add testing via Travis, implement the cloning, uploading, and writing, and include a template notebook (or at least a document) showcasing Google Drive support.

### CODE REVIEW

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good so far. This is a fair amount of complexity to add to the repo, but given how many people store their data in google drive I think it will probably be worth it.

File: ark/utils/google_drive_utils.py
Comment (alex-l-kong) [inline ark/utils/google_drive_utils.py:330]: Can we put this code in a separate function? Seems like similar logic is repeated in `lsfiles`.

**Review by alex-l-kong — CHANGES_REQUESTED** [review]
Looks good so far. Just one thing I've noticed so far.

**Comment by ackagel** [comment]
still some more testing functions to write, but everything is implemented, and the mocking setup is complete.  An example of how it's used 'per-test' is on lines 336-340 of `google_drive_utils_test.py`

**Review by alex-l-kong — COMMENTED** [review]
Just one comment from me for now

File: ark/utils/google_drive_utils.py
Comment (bcollica) [reply ark/utils/google_drive_utils.py:39]: +1

File: ark/utils/google_drive_utils.py
Comment (bcollica) [inline ark/utils/google_drive_utils.py:314]: Might be helpful to provide some other error messages if the src path does not exist or if there are other failures.  There seem to be sufficient warnings for the overwrite cases and also for successful uploads, but it seems like this would return nothing if I incorrectly specified the local path containing the data I want to upload?

File: ark/utils/google_drive_utils.py
Comment (ackagel) [reply ark/utils/google_drive_utils.py:39]: We can do that.  The salt will technically still be viewable to anyone who digs around for it, so I figured putting it directly in the utils file was fine.  Unless we're willing to have a whole server dedicated to hiding the api token, salt, etc, there's not a whole lot to do about hiding how the sausage is made (sausage being token decryption).

**Review by ngreenwald — CHANGES_REQUESTED** [review]
1. Have you gone through and tested all of this with an actual google drive file system? The mocked tests look fairly comprehensive, but there will inevitably be stuff that they miss. 
2. You should update the Jupyter Notebook with a description of how to specify a google drive file, the authentication process, etc.

**Comment by ackagel** [comment]
I know this is a massive PR, so to clarify, I mostly need feedback for the example google drive notebook (stuff like implementation, understandability, etc.).

Regardless, no more massive PR's from me after this one!

**Comment by ngreenwald** [comment]
Is the idea that we'd have a separate version of each notebook that shows how google drive paths can be supported? And that each person who wants to use the google drive path would need to get the passcode from you?

**Comment by ackagel** [comment]
I was thinking one demo notebook that exhaustively demonstrates the general interoperability of GoogleDrivePaths and shows how to resolve some inevitable conflicts (like file writes hidden within a function call ala `io.imsave` or `df.to_csv`).

**Comment by ackagel** [comment]
Passcode wise, yeah they'll have to reach out to anyone who knows the passcode.  I can put my email address in there for now, but I'm open to suggestions.

**Review by bcollica — COMMENTED** [review]
A few nitpicks, but overall very impressive of course.  My main issue is with the notebook example.  I think having more distinct sections based on functionality/tasks and also clearer descriptions of what is being done (and maybe why) would be helpful, especially for a new user looking to this notebook for guidance.

File: templates/Segment_Image_Data_GoogleDriveExample.ipynb
Comment (ackagel) [reply templates/Segment_Image_Data_GoogleDriveExample.ipynb:38]: yeah haha, that's filler.  I'll throw in my email/slack/etc in there at somepoint.

File: templates/Segment_Image_Data_GoogleDriveExample.ipynb
Comment (ackagel) [reply templates/Segment_Image_Data_GoogleDriveExample.ipynb:73]: They're all google drive paths for demonstration purposes.  

`deepcell_input_dir` and a few others should be local in practice since they hold 'mid-processing' stages of data, so if the demonstration is just confusing, I can change that to be more practical.

File: templates/Segment_Image_Data_GoogleDriveExample.ipynb
Comment (ackagel) [reply templates/Segment_Image_Data_GoogleDriveExample.ipynb:274]: I think that's a vestigial H2 from an older version of the notebooks.  I'll H3 that and change Deepcell to Mesmer since I think that's the official ✨ _branding_ ✨  these days

File: templates/Segment_Image_Data_GoogleDriveExample.ipynb
Comment (ackagel) [reply templates/Segment_Image_Data_GoogleDriveExample.ipynb:399]: I could rework the markdown to make the code block more condensed.

If I put the example in a live code cell, some people will run it, be upset when it doesn't work, and maybe post an issue about it, regardless of how bold or red-lettered the message explicitly stating not to run the example is.  

_This is the truth I've divined through my pain_

File: templates/Segment_Image_Data_GoogleDriveExample.ipynb
Comment (ackagel) [reply templates/Segment_Image_Data_GoogleDriveExample.ipynb:425]: Yeah this is more or less what I was antsy about.  There's three work arounds that I can think of to a lambda function.

 1. Assume all write out api's have the syntax `library.write_out(path, *args, **kwargs)` where the filepath is always the first argument, so the lambda can be pretty easily constructed behind the scenes.  Surprisingly haven't found a concrete example of this not being the case, but that could always crop up and be a little nasty.
 2. Assume the user has knowledge of the required function argument names and use that to construct the lambda function.  This looks something like `drive_write_out(library.write_out, path_argument=path, **kwargs)` or behind the scenes:
```
# inside of drive_write_out(func, path_argument, **kwargs)
def write_out_lambda(x):
    kwargs.update({path_argument: x})
    func(**kwargs)
```
 3. Write individual wrapper functions for common file write out api's, but leave lambda functionality as is for more niche use cases.  This is probably what I'd go with, but I'm super open to y'all's thoughts on this.

File: ark/utils/deepcell_service_utils.py
Comment (ackagel) [reply ark/utils/deepcell_service_utils.py:193]: 'responce' is my most common misspelling, and it continues to haunt me 👻 .  I can't get the incorrect mnemonic res-'Ponce de Leon' out of my head.

**Review by ngreenwald — CHANGES_REQUESTED** [review]
I think it makes sense to transition from a notebook to pure documentation. This will allow you to cut out all the code that is there to make it a functional notebook, and focus only on the key features you want to explain

**Review by alex-l-kong — COMMENTED** [review]
Nothing else to add on top of the other comments.

**Review by ngreenwald — COMMENTED** [review]
I looked through all the coverage decreases. None of them will affect anyone using the default pipeline, since the logic to default to local paths is getting tested with existing tests. Have you gone through and verified that the whole segmentation notebook works as is? If so I think it's fine to merge in and bugfix stuff once people find corner cases we didn't think of

**Review by bcollica — COMMENTED** [review]
looks great

**Comment by ackagel** [comment]
Yep; the segmentation notebook still works out of the box with local paths, as well as google drive paths (with relevant tweaks).  If we get some consistent level of bug reports, I can rework some tests to cover google drive path support more thoroughly.

Just need an approval for merging in

**Pull request merged**
